casanovo/denovo/model.py

```diff
@@ -176,12 +193,592 @@ def forward(
+         scores : torch.Tensor of shape (n_spectra*  n_beams, max_length, n_amino_acids)
+            Output scores of the model.
+        idx : int
+            Index to be considered in the current decoding step
+
+        Returns
+        -------
+        finished_beams_idx : List[int]
+            Indices of all finished beams on tokens tensor
+        cache_idx_dict : Dict[int, int]
+            Next available tensor index to cache peptides for each spectrum.
+        cache_seq_dict : Dict[int,  Set[str]]
+            Set of decoded peptides for each spectrum.
+        cache_score_dict : Dict[int, List[List[Tuple[float, int]]]
+            Confidence score for each decoded peptide, separated as precursor m/z fitting vs not, for each spectrum.
+
+        """
+        beam = self.n_beams
+        # Store finished beams in the cache
+        if len(finished_beams_idx) > 0:
+            for i in finished_beams_idx:
+                spec_idx = i // beam  # find the starting index of the spectrum
+                insert_idx = cache_idx_dict[spec_idx]
+
+                # Only add beams with matching mass to the cache
+                prec_mz = precursors[i, 2].item()
+                aa_list = [
+                    self.decoder._idx2aa.get(idx_aa.item(), "")
+                    for idx_aa in tokens[i][:idx]
+                ]
+
+                # Check if dummy token predicted
+                if "" not in aa_list:
+                    # Remove stop token as needed, i.e. if mass filter induced termination
+                    if "$" in aa_list:
+                        aa_list.remove("$")
+
+                    pep_str = "".join(aa_list)
+                    pred_mz = self.peptide_mass_calculator.mass(
+                        aa_list, charge=precursors[i, 1].item()
+                    )
+                    delta_mass_ppm = (pred_mz - prec_mz) / prec_mz * 10**6
+                    isPrecursorFit = (
+                        abs(delta_mass_ppm) <= self.precursor_mass_tol
+                    )
+
+                    # Check if predicted seq already in cache
+                    if pep_str not in cache_seq_dict[spec_idx]:
+                        # Directly cache if we don't already have k peptides cached
+                        if insert_idx < (spec_idx + 1) * beam:
+                            cache_tokens[insert_idx, :] = tokens[i, :]
+                            cache_scores[insert_idx, :, :] = scores[i, :, :]
+                            cache_idx_dict[spec_idx] += 1  # move the pointer
+                            cache_seq_dict[spec_idx].add(pep_str)
+                            smx = self.softmax(scores)[i, :, :]
+
+                            aa_scores = [
+                                smx[idx][self.decoder._aa2idx[aa]].item()
+                                for idx, aa in enumerate(aa_list)
+                            ]
+                            pep_score = np.mean(aa_scores)
```

Comment:
**suggestion:** It might be useful to extract this behavior to a AA-scores-to-peptide-scores function, so that if we want to change it, we can easily do it in a single location.

Reply:
Extracted to `_aa_to_pep_score()`